provisioning/device/src/CertificateInstaller.cs

```diff
@@ -1,68 +1,31 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
-using Microsoft.Azure.Devices.Shared;
-using System;
-using System.Collections.Generic;
 using System.Security.Cryptography.X509Certificates;
+using Microsoft.Azure.Devices.Shared;
 
 namespace Microsoft.Azure.Devices.Provisioning.Client
 {
     internal static class CertificateInstaller
     {
-        private static readonly HashSet<string> s_installedCertificates = new HashSet<string>();
-        private static readonly object s_lock = new object();
-
-        static CertificateInstaller()
-        {
-            try
-            {
-                using (var store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser))
-                {
-                    store.Open(OpenFlags.ReadOnly);
-                    foreach (X509Certificate2 certificate in store.Certificates)
-                    {
-                        s_installedCertificates.Add(certificate.Thumbprint);
-                    }
-                }
-            }
-            catch (Exception ex)
-            {
-                if (Logging.IsEnabled)
-                {
-                    Logging.Error(
-                     null,
-                     $"{nameof(CertificateInstaller)} failed to read store: {ex}.");
-                }
-            }
-        }
-
         public static void EnsureChainIsInstalled(X509Certificate2Collection certificates)
         {
             if (certificates == null)
             {
                 return;
             }
 
-            lock (s_lock)
+            using var store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser);
+            store.Open(OpenFlags.ReadWrite);
+
+            foreach (X509Certificate2 certificate in certificates)
             {
-                using (var store = new X509Store(StoreName.CertificateAuthority, StoreLocation.CurrentUser))
+                X509Certificate2Collection results = store.Certificates.Find(X509FindType.FindByThumbprint, certificate.Thumbprint, false);
+                if (results.Count == 0)
                 {
-                    store.Open(OpenFlags.ReadWrite);
-
-                    foreach (X509Certificate2 certificate in certificates)
-                    {
-                        if (!s_installedCertificates.Contains(certificate.Thumbprint))
-                        {
-                            if (Logging.IsEnabled)
-                            {
-                                Logging.Info(null, $"{nameof(CertificateInstaller)} adding {certificate.Thumbprint}");
-                            }
+                    Logging.Info(null, $"{nameof(CertificateInstaller)} adding {certificate.Thumbprint}");
```

Comment:
Perf: (I copied logging from .NET Core.)
If (Logging.IsEnabled) is preventing the string expansion (slow) from happening if the log is not enabled. Removing it will perform the string formatter _always_.

The recommendation is to add the check whenever the logging line has parameters that end up taking CPU cycles (all except passing already evaluated primitive types). 

E.g. I think Log.Associate below doesn't require the if check (it does save one if branch but I don't see that as important as string formatting for example).